### PR TITLE
feat: convert year_dropdown.jsx and year_dropdown_options.jsx into ts

### PR DESCRIPTION
**Pull request by yuki0410-dev**
## Description
**Linked issue**: #4700

**Changes** `year_dropdown.jsx and year_dropdown_options.jsx` was migrated to ts and test

## Contribution checklist
* [x]  I have followed the [contributing guidelines](https://github.com/Hacker0x01/react-datepicker/blob/main/CONTRIBUTING.md).
* [x]  I have added sufficient test coverage for my changes.
* [x]  I have formatted my code with Prettier and checked for linting issues with ESLint for code readability.

### CODE REVIEW

File: src/year_dropdown_options.tsx [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
-import PropTypes from "prop-types";
 import { clsx } from "clsx";
 import { getYear } from "./date_utils";
 
-function generateYears(year, noOfYear, minDate, maxDate) {
-  const list = [];
+function generateYears(
+  year: number,
+  noOfYear: number,
+  minDate: Date | undefined,
```
Comment (mirus-ua) [inline src/year_dropdown_options.tsx:8]: IMO, `?` is preferable to mark the prop as optional rather than explicitly using `undefined` notation

Comment (yuki0410-dev) [reply src/year_dropdown_options.tsx]: @mirus-ua
Thanks for the advice. Fixed it.

File: test/calendar_test.test.js [hunk 2]
```diff
@@ -1214,7 +1214,11 @@ describe("Calendar", () => {
       const select = calendar
         .querySelector(".react-datepicker__year-dropdown-container")
         .querySelector("select");
-      fireEvent.change(select);
+      fireEvent.change(select, {
+        target: {
```
Comment (mirus-ua) [inline test/calendar_test.test.js:1218]: Is it a breaking change that you modified a test? If yes, I would like to document it in the OG issue https://github.com/Hacker0x01/react-datepicker/issues/4700

**Comment by mirus-ua** [comment]
> @mirus-ua (@martijnrusschen ) The Codecov Report has indicated a drop in coverage, but as you can see from the diff, I only added the TS type and made no changes to the JS implementation part. Should I still prevent the coverage rate from dropping in this case?

No, this is expected

File: src/year_dropdown.tsx [hunk 3]
```diff
@@ -81,7 +89,7 @@ export default class YearDropdown extends React.Component {
     />
   );
 
-  renderScrollMode = () => {
+  renderScrollMode = (): JSX.Element[] => {
     const { dropdownVisible } = this.state;
     let result = [this.renderReadView(!dropdownVisible)];
```
[the remark this answers is not in the document]
Comment (yuki0410-dev) [reply src/year_dropdown.tsx]: Fixed it.

**Comment by yuki0410-dev** [comment]
@martijnrusschen
The changes you want to use in tsing another file are included in this PR.
Please merge if it is OK.

**Pull request merged**